### PR TITLE
[Bugfix] - Additional Feedback consistency

### LINKED ISSUE #1708: Additional Feedback not displayed on team exercises for students
#### To Reproduce
1. Create a team exercise
2. Participate with a student
3. Assess this with additional feedback
4. Check as student and see that you have no feedback

#### Expected behavior
The feedback gets presented

#### Additional context
It seems like the content of the feedback doesn't event get passed in the call.

### PR DESCRIPTION
**Pull request by FuchsDominik**
Fixes #1708

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context

We handled additional feedback (or at is was called before, unreferenced feedback) differently for every exercise type. This should be consistent.

### Description

- Added a additonal-feedback component which shows the additional feedback done by the assessor and this component is used by every exercise type

### Steps for Testing

1. Log in to Artemis
2. For text-/modeling-/file-upload- and programming exercises do:
2.1. Create an exercise and participate
2.2. Assess the submission (once with additional feedback (one or multiple) and once without additional feedback)
2.3. With the student, go to the submission, below the submission, the additional feedback should be shown in the same way for every exercise or should be not shown if no additional feedback provided

Aspects to also review:
-  File upload exercises have some wording adaptions, see screenshot
- For modeling exercises, the additional feedback should not be shown in the feedback list next to the editor (At the beginning)

### Screenshots

Text Exercise:
Screenshot 2021-03-03 at 14 06 49

Modeling Exercise:
Screenshot 2021-03-03 at 14 06 34

File-Upload Exercise:
Screenshot 2021-03-03 at 14 06 56

Programming Exercise:
Screenshot 2021-03-03 at 22 07 27

### CODE REVIEW

**Review by JannisBittlmayer — DISMISSED** [review]
Tested with @FlorianStang, works as intended

**Review by FlorianStang — APPROVED** [review]
Tested on TS1 with @JannisBittlmayer, everything worked as intended.

**Review by airelawaleria — APPROVED** [review]

**Review by EneaGore — DISMISSED** [review]
Looks cleaner and seems to work well.

**Review by JohannesWeiss — DISMISSED** [review]
Tested on ts1, works mostly as intended.

During testing, the modelling editor in the results view behaved very strange: 


https://user-images.githubusercontent.com/37304775/110007474-ff40f180-7d1a-11eb-95ce-74176c3f5b42.mov

First, the diagram is shown for a brief moment, before it (sometimes completely) disappears. This seems to be related to the dimensions of the window, since after a resize it re-appears. Reloading the window with the new dimensions reproduces the issue. Since this seems like an edge case (in Safari only as far as I could test) and is probably unrelated to this PR, I'll approve nevertheless ;)

**Review by sebastiantpeter — APPROVED** [review]
Looks good on ts1.

**Comment by FuchsDominik** [comment]
> Tested on ts1, works mostly as intended.
> 
> During testing, the modelling editor in the results view behaved very strange:
> 
>  Screen.Recording.2021-03-04.at.18.45.50.mov 
> First, the diagram is shown for a brief moment, before it (sometimes completely) disappears. This seems to be related to the dimensions of the window, since after a resize it re-appears. Reloading the window with the new dimensions reproduces the issue. Since this seems like an edge case (in Safari only as far as I could test) and is probably unrelated to this PR, I'll approve nevertheless ;)

@JohannesWeiss This is indeed weird behaviour. If this happens on a regular basis, you could create an issue for that. I did not touch the editor component, though, so like you said not related to this PR :)

File: src/main/webapp/app/exercises/file-upload/participate/file-upload-submission.component.html [hunk 1]
```diff
@@ -50,8 +50,16 @@ <h6>
         </div>
         <div *ngIf="result && result.feedbacks && result.feedbacks.length > 0">
             <br />
-            <h5><span jhiTranslate="artemisApp.fileUploadExercise.assessedSubmission">Your assessed submission</span>:</h5>
-            <jhi-file-upload-result [result]="result"></jhi-file-upload-result>
+            <h5><span jhiTranslate="artemisApp.fileUploadExercise.assessedSubmission">Your Assessment</span>:</h5>
+            <div class="row">
+                <jhi-additional-feedback
+                    *ngIf="unreferencedFeedback && unreferencedFeedback.length != 0"
```
Comment (fde312) [inline src/main/webapp/app/exercises/file-upload/participate/file-upload-submission.component.html:56]: I think you can simply that and I would use > 0 instead != 0 :) 

```suggestion
                    *ngIf="unreferencedFeedback?.length > 0"
```

File: src/main/webapp/app/exercises/file-upload/participate/file-upload-submission.component.ts [hunk 2]
```diff
@@ -186,6 +187,15 @@ export class FileUploadSubmissionComponent implements OnInit, ComponentCanDeacti
         }
     }
 
+    /**
+     * Find "Unreferenced Feedback" item for Result, if it exists.
+     */
+    get unreferencedFeedback(): Feedback[] | undefined {
+        if (this.result && this.result.feedbacks && Array.isArray(this.result.feedbacks)) {
```
Comment (fde312) [inline src/main/webapp/app/exercises/file-upload/participate/file-upload-submission.component.ts:194]: Why are you using `Array.isArray(this.result.feedback)` should be an array or undefined anyways or not?

File: src/main/webapp/app/exercises/modeling/participate/modeling-submission.component.html [hunk 3]
```diff
@@ -103,10 +102,19 @@ <h4 jhiTranslate="artemisApp.modelingSubmission.assessment.title">Assessment</h4
                 </div>
             </div>
         </jhi-fullscreen>
-        <jhi-complaint-interactions *ngIf="modelingExercise" [exercise]="modelingExercise" [participation]="participation" [result]="result"> </jhi-complaint-interactions>
+        <div class="row">
+            <jhi-additional-feedback
+                *ngIf="unreferencedFeedback && unreferencedFeedback.length != 0"
```
Comment (fde312) [inline src/main/webapp/app/exercises/modeling/participate/modeling-submission.component.html:107]: see as for file upload

File: src/main/webapp/app/exercises/file-upload/participate/file-upload-submission.component.ts (hunk 2 above)
Comment (fde312) [reply src/main/webapp/app/exercises/file-upload/participate/file-upload-submission.component.ts:194]: This is also for the other exercise types.

File: src/main/webapp/app/exercises/file-upload/participate/file-upload-submission.component.html (hunk 1 above)
Comment (fde312) [reply src/main/webapp/app/exercises/file-upload/participate/file-upload-submission.component.html:56]: This is also for the other exercise types

File: src/main/webapp/app/exercises/modeling/participate/modeling-submission.component.ts [hunk 4]
```diff
@@ -401,6 +400,24 @@ export class ModelingSubmissionComponent implements OnInit, OnDestroy, Component
...
+            return this.assessmentResult.feedbacks.filter((feedbackElement) => feedbackElement.reference == undefined && feedbackElement.type === FeedbackType.MANUAL_UNREFERENCED);
+        }
+    }
+
+    /**
+     * Find "Referenced Feedback" item for Result, if it exists.
+     */
+    get referencedFeedback(): Feedback[] | undefined {
+        if (this.assessmentResult && this.assessmentResult.feedbacks && Array.isArray(this.assessmentResult.feedbacks)) {
+            return this.assessmentResult.feedbacks.filter((feedbackElement) => feedbackElement.reference !== undefined);
```
Comment (fde312) [inline src/main/webapp/app/exercises/modeling/participate/modeling-submission.component.ts:417]:
```suggestion
            return this.assessmentResult.feedbacks.filter((feedbackElement) => feedbackElement.reference != undefined);
```

File: src/main/webapp/app/exercises/modeling/participate/modeling-submission.component.ts [hunk 5]
```diff
@@ -401,6 +400,24 @@ export class ModelingSubmissionComponent implements OnInit, OnDestroy, Component
...
+        if (this.assessmentResult && this.assessmentResult.feedbacks && Array.isArray(this.assessmentResult.feedbacks)) {
+            return this.assessmentResult.feedbacks.filter((feedbackElement) => feedbackElement.reference == undefined && feedbackElement.type === FeedbackType.MANUAL_UNREFERENCED);
+        }
+    }
+
+    /**
+     * Find "Referenced Feedback" item for Result, if it exists.
+     */
+    get referencedFeedback(): Feedback[] | undefined {
+        if (this.assessmentResult && this.assessmentResult.feedbacks && Array.isArray(this.assessmentResult.feedbacks)) {
```
Comment (fde312) [inline src/main/webapp/app/exercises/modeling/participate/modeling-submission.component.ts:416]: same as for unreferenced feedback `Array.isArray`

File: src/main/webapp/app/exercises/text/participate/text-editor.component.html [hunk 6]
```diff
@@ -72,27 +72,14 @@
...
-                            [class.negative-feedback]="feedback?.credits < 0"
-                            [class.neutral-feedback]="feedback?.credits === 0"
-                        >
-                            <strong>
-                                <span> {{ feedback.credits + ' Points: ' }} </span>
-                            </strong>
-                            <span> {{ feedback.detailText }} </span>
-                        </div>
-                    </div>
+                    <!-- TODO use following component also for modeling, text, file-upload and programming -->
```
Comment (fde312) [inline src/main/webapp/app/exercises/text/participate/text-editor.component.html:75]: remove todo

**Review by fde312 — DISMISSED** [review]
Code looks good to me, I did not test it on the test server yet. I have just some small remarks (see my comments).

Other than that really cool enhancement 👍

File: src/main/webapp/app/exercises/modeling/participate/modeling-submission.component.ts (hunk 4 above)
Comment (FuchsDominik) [reply src/main/webapp/app/exercises/modeling/participate/modeling-submission.component.ts:417]: Does it make a difference here? I wouldn't say type-checking is unnecessary. Maybe someone uses the reference "undefined" :P

File: src/main/webapp/app/exercises/text/participate/text-editor.component.html (hunk 6 above)
Comment (FuchsDominik) [reply src/main/webapp/app/exercises/text/participate/text-editor.component.html:75]: 👍

File: src/main/webapp/app/exercises/file-upload/participate/file-upload-submission.component.html (hunk 1 above)
Comment (FuchsDominik) [reply src/main/webapp/app/exercises/file-upload/participate/file-upload-submission.component.html:56]: Sounds good :)

File: src/main/webapp/app/exercises/file-upload/participate/file-upload-submission.component.ts (hunk 2 above)
Comment (FuchsDominik) [reply src/main/webapp/app/exercises/file-upload/participate/file-upload-submission.component.ts:194]: I just reused this method. Don't know if the developing person had something particular in mind, but since feedback can indeed only be an array or null this check could probably be removed

File: src/main/webapp/app/exercises/modeling/participate/modeling-submission.component.ts (hunk 5 above)
Comment (FuchsDominik) [reply src/main/webapp/app/exercises/modeling/participate/modeling-submission.component.ts:416]: 👍

File: src/main/webapp/app/shared/additional-feedback/additional-feedback.component.ts [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+@Component({
+    selector: 'jhi-additional-feedback',
+    templateUrl: './additional-feedback.component.html',
+    styleUrls: ['./additional-feedback.component.scss'],
+})
+export class AdditionalFeedbackComponent {
+    @Input()
+    additionalFeedback: Feedback[];
+    @Input()
+    additional: boolean;
```
Comment (FrankeLukas) [inline src/main/webapp/app/shared/additional-feedback/additional-feedback.component.ts:13]: This boolean is a bit confusing given that you already provided *addition* feedback in the other input. How about we call the first input just `feedback: Feedback[]`? Then the meaning of this boolean is a bit more obvious imo.

File: src/main/webapp/app/exercises/modeling/participate/modeling-submission.component.ts (hunk 4 above)
Comment (FrankeLukas) [reply src/main/webapp/app/exercises/modeling/participate/modeling-submission.component.ts:417]: It does make a difference, using just `!=` also accounts for the variable being `null`.

**Review by FrankeLukas — DISMISSED** [review]

File: src/main/webapp/app/exercises/modeling/participate/modeling-submission.component.ts (hunk 4 above)
Comment (FuchsDominik) [reply src/main/webapp/app/exercises/modeling/participate/modeling-submission.component.ts]: Alright, i am giving in :D

File: src/main/webapp/app/shared/additional-feedback/additional-feedback.component.ts (hunk 7 above)
Comment (FuchsDominik) [reply src/main/webapp/app/shared/additional-feedback/additional-feedback.component.ts:13]: 👍

**Review by nicolasruscher — DISMISSED** [review]
Code looks good to me!  But I did not test it

**Review by tiffca — APPROVED** [review]
Tested successfully

File: src/main/webapp/app/exercises/file-upload/participate/file-upload-submission.component.ts [hunk 8]
```diff
@@ -186,6 +187,15 @@ export class FileUploadSubmissionComponent implements OnInit, ComponentCanDeacti
         }
     }
 
+    /**
+     * Find "Unreferenced Feedback" item for Result, if it exists.
+     */
+    get unreferencedFeedback(): Feedback[] | undefined {
+        if (this.result && this.result.feedbacks) {
+            return this.result.feedbacks.filter((feedbackElement) => feedbackElement.reference == undefined && feedbackElement.type === FeedbackType.MANUAL_UNREFERENCED);
```
Comment (SimonEntholzer) [inline src/main/webapp/app/exercises/file-upload/participate/file-upload-submission.component.ts:195]: here you could use `!feedbackElement.reference`

**Review by SimonEntholzer — DISMISSED** [review]
Tested on ts3, everything worked fine :D the code looks good too
I really like how clean it looks now

File: src/main/webapp/app/exercises/file-upload/participate/file-upload-submission.component.ts (hunk 8 above)
Comment (FuchsDominik) [reply src/main/webapp/app/exercises/file-upload/participate/file-upload-submission.component.ts:195]: Probably really just a matter of taste but I would prefer to see the == undefined here :P

**Review by derLalla — CHANGES_REQUESTED** [review]
Nice code changes. I have left some small remarks.

**Review by ivanchimeno — DISMISSED** [review]
The code looks good, I approve :)

File: src/main/webapp/app/exercises/programming/participate/code-editor-student-container.component.ts [hunk 9]
```diff
@@ -162,4 +162,13 @@ export class CodeEditorStudentContainerComponent implements OnInit, OnDestroy, C
         }
         this.hasTutorAssessment = isManualResult && hasTutorFeedback;
     }
+
+    /**
+     * Find "Unreferenced Feedback" item for Result, if it exists.
+     */
+    get unreferencedFeedback(): Feedback[] | undefined {
```
[the remark this answers is not in the document]
Comment (fde312) [reply src/main/webapp/app/exercises/programming/participate/code-editor-student-container.component.ts:169]: Good idea, than you just need to change it once and it works for all exercise types! :)

**Review by fde312 — COMMENTED** [review]
For me everything worked fine. I agree with @derLalla about moving it to a util service :)

File: src/main/webapp/app/exercises/modeling/participate/modeling-submission.component.ts [hunk 10]
```diff
@@ -401,6 +400,24 @@ export class ModelingSubmissionComponent implements OnInit, OnDestroy, Component
...
+    }
+
+    /**
+     * Find "Referenced Feedback" item for Result, if it exists.
+     */
+    get referencedFeedback(): Feedback[] | undefined {
+        if (this.assessmentResult && this.assessmentResult.feedbacks) {
+            return this.assessmentResult.feedbacks.filter((feedbackElement) => feedbackElement.reference != undefined);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (FuchsDominik) [reply src/main/webapp/app/exercises/modeling/participate/modeling-submission.component.ts:419]: 👍

File: src/main/webapp/app/exercises/programming/participate/code-editor-student-container.component.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (FuchsDominik) [reply src/main/webapp/app/exercises/programming/participate/code-editor-student-container.component.ts:170]: c01428c

File: src/main/webapp/app/exercises/modeling/participate/modeling-submission.component.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (FuchsDominik) [reply src/main/webapp/app/exercises/modeling/participate/modeling-submission.component.ts]: Sounds good :)

File: src/main/webapp/app/exercises/modeling/participate/modeling-submission.component.ts [hunk 11]
```diff
@@ -401,6 +400,24 @@ export class ModelingSubmissionComponent implements OnInit, OnDestroy, Component
...
     }
 
+    /**
+     * Find "Unreferenced Feedback" item for Result, if it exists.
+     */
+    get unreferencedFeedback(): Feedback[] | undefined {
+        if (this.assessmentResult && this.assessmentResult.feedbacks) {
+            return this.assessmentResult.feedbacks.filter((feedbackElement) => feedbackElement.reference == undefined && feedbackElement.type === FeedbackType.MANUAL_UNREFERENCED);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (FuchsDominik) [reply src/main/webapp/app/exercises/modeling/participate/modeling-submission.component.ts]: Sounds good, c01428c :)

File: src/main/webapp/app/exercises/file-upload/participate/file-upload-submission.component.ts (hunk 8 above)
Comment (FuchsDominik) [reply src/main/webapp/app/exercises/file-upload/participate/file-upload-submission.component.ts]: 👍  c01428c

**Review by FrankeLukas — APPROVED** [review]
Lgtm now and works as advertised.

File: src/main/webapp/app/exercises/modeling/participate/modeling-submission.component.ts [hunk 12]
```diff
@@ -6,7 +6,7 @@ import { NgbModal } from '@ng-bootstrap/ng-bootstrap';
 import { TranslateService } from '@ngx-translate/core';
 import { JhiWebsocketService } from 'app/core/websocket/websocket.service';
 import { ComplaintType } from 'app/entities/complaint.model';
-import { Feedback } from 'app/entities/feedback.model';
+import { Feedback, FeedbackType } from 'app/entities/feedback.model';
```
Comment (FrankeLukas) [inline src/main/webapp/app/exercises/modeling/participate/modeling-submission.component.ts:9]: Unused import.

File: src/main/webapp/app/exercises/text/participate/text-editor.component.ts [hunk 13]
```diff
@@ -168,17 +169,10 @@ export class TextEditorComponent implements OnInit, OnDestroy, ComponentCanDeact
     }
 
     /**
-     * Find "Unreferenced Feedback" item for Result, if it exists.
+     * Check whether or not a result exists and if, returns the unreferenced feedback of it
      */
-    get unreferencedFeedback(): Feedback[] | null {
-        if (this.result && this.result.feedbacks && Array.isArray(this.result.feedbacks)) {
-            const feedbackWithoutReference = this.result.feedbacks.filter(
-                (feedbackElement) => feedbackElement.reference == undefined && feedbackElement.type === FeedbackType.MANUAL_UNREFERENCED,
```
Comment (FrankeLukas) [inline src/main/webapp/app/exercises/text/participate/text-editor.component.ts:176]: This removes the last usage of `FeedbackType`, so we get an unused import warning in this file too.

**Review by FrankeLukas — CHANGES_REQUESTED** [review]

**Review by FrankeLukas — APPROVED** [review]
Resolved the style issues

**Review by fde312 — APPROVED** [review]
Retested everything, works fine! 👍
